hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
+                vacant_entry.insert(root);
+                for &v in preds.get(&u).into_iter().flatten() {
+                    assign(preds, v, root, components);
+                }
             }
+        }
 
-            (node_subgraph, subgraph_nodes)
-        };
-
-        // Get data on handoff src and dst subgraphs.
-        let mut subgraph_recv_handoffs: SecondaryMap<GraphSubgraphId, Vec<GraphNodeId>> =
-            subgraph_nodes
-                .keys()
-                .map(|k| (k, Default::default()))
-                .collect();
-        let mut subgraph_send_handoffs = subgraph_recv_handoffs.clone();
-        for edge in iter_edges(&new_succs) {
-            let ((src, _), (dst, _)) = edge;
-            let (src_node, dst_node) = (&flat_graph.nodes[src], &flat_graph.nodes[dst]);
-            match (src_node, dst_node) {
-                (Node::Operator(_), Node::Operator(_)) => {}
-                (Node::Operator(_), Node::Handoff) => {
-                    subgraph_send_handoffs[node_subgraph[src]].push(dst);
-                }
-                (Node::Handoff, Node::Operator(_)) => {
-                    subgraph_recv_handoffs[node_subgraph[dst]].push(src);
+        let mut components = Default::default();
+        for sg in stack.into_iter().rev() {
+            assign(&subgraph_preds, sg, sg, &mut components);
+        }
+
+        components
+    };
+
+    let topo_sort_order = {
+        // Condensed each SCC into a single node for toposort.
+        let condensed_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = subgraph_preds
+            .iter()
+            .map(|(u, preds)| (scc[u], preds.iter().map(|v| scc[v]).collect()))
+            .collect();
+
+        graph_algorithms::topo_sort(subgraph_nodes.keys(), |v| {
+            condensed_preds.get(&v).into_iter().flatten().cloned()
+        })
+    };
+
+    let mut subgraph_stratum: SecondaryMap<GraphSubgraphId, usize> = topo_sort_order
+        .into_iter()
+        .enumerate()
+        .map(|(n, sg_id)| (sg_id, n))
+        .collect();
+
+    let max_stratum = subgraph_stratum.values().cloned().max().unwrap_or(0) + 1;
+    for &(src, dst, dst_idx, input_barrier) in barrier_crossers.iter() {
+        let src_sg = node_subgraph[src];
+        let dst_sg = node_subgraph[dst];
+        let src_stratum = subgraph_stratum[src_sg];
+        let dst_stratum = subgraph_stratum[dst_sg];
+        match input_barrier {
+            InputBarrier::None => unreachable!("No barrier"),
+            InputBarrier::Stratum => {
+                // Any negative edges which go onto the same or previous stratum are bad.
+                // Indicates an unbroken negative cycle.
+                if dst_stratum <= src_stratum {
+                    dst_idx
+                        .span()
+                        .unwrap()
+                        .error("Connection creates a negative cycle which must be broken with a `next_epoch()` operator.")
```

Comment:
would it be possible for the error message to tell the user where to look in the graph for the semantic bug?

Reply:
Shows it on the negative port edges